core/src/layers/throttle.rs

```diff
@@ -0,0 +1,332 @@
+    }
+
+    fn blocking_read(&self, path: &str, args: OpRead) -> Result<(RpRead, Self::BlockingReader)> {
+        let limiter = self.rate_limiter.clone();
+
+        self.inner
+            .blocking_read(path, args)
+            .map(|(rp, r)| (rp, ThrottleWrapper::new(r, limiter)))
+    }
+
+    fn blocking_write(&self, path: &str, args: OpWrite) -> Result<(RpWrite, Self::BlockingWriter)> {
+        let limiter = self.rate_limiter.clone();
+
+        self.inner
+            .blocking_write(path, args)
+            .map(|(rp, w)| (rp, ThrottleWrapper::new(w, limiter)))
+    }
+
+    fn blocking_list(&self, path: &str, args: OpList) -> Result<(RpList, Self::BlockingPager)> {
+        self.inner.blocking_list(path, args)
+    }
+}
+
+pub struct ThrottleWrapper<R> {
+    inner: R,
+    limiter: SharedRateLimiter,
+}
+
+impl<R> ThrottleWrapper<R> {
+    pub fn new(inner: R, rate_limiter: SharedRateLimiter) -> Self {
+        Self {
+            inner,
+            limiter: rate_limiter,
+        }
+    }
+}
+
+impl<R: oio::Read> oio::Read for ThrottleWrapper<R> {
+    fn poll_read(&mut self, cx: &mut Context<'_>, buf: &mut [u8]) -> Poll<Result<usize>> {
+        // TODO: How can we handle buffer reads with a limiter?
+        self.inner.poll_read(cx, buf)
+    }
+
+    fn poll_seek(&mut self, cx: &mut Context<'_>, pos: SeekFrom) -> Poll<Result<u64>> {
+        self.inner.poll_seek(cx, pos)
+    }
+
+    fn poll_next(&mut self, cx: &mut Context<'_>) -> Poll<Option<Result<Bytes>>> {
+        self.inner.poll_next(cx)
+    }
+}
+
+impl<R: oio::BlockingRead> oio::BlockingRead for ThrottleWrapper<R> {
+    fn read(&mut self, buf: &mut [u8]) -> Result<usize> {
+        // TODO: How can we handle buffer reads with a limiter?
+        self.inner.read(buf)
+    }
+
+    fn seek(&mut self, pos: SeekFrom) -> Result<u64> {
+        self.inner.seek(pos)
+    }
+
+    fn next(&mut self) -> Option<Result<Bytes>> {
+        self.inner.next()
+    }
+}
+
+#[async_trait]
+impl<R: oio::Write> oio::Write for ThrottleWrapper<R> {
+    async fn write(&mut self, mut bs: Bytes) -> Result<()> {
+        // Note: for self.inner.write(bs), it's possible that the given bs length is less than the total content length.
+        // Users will call write multiple times to write the whole data.
+
+        let mut buf_length = NonZeroU32::new(bs.len() as u32).unwrap();
+
+        while !bs.is_empty() {
+            match self.limiter.check_n(buf_length) {
+                Ok(_) => self.inner.write(bs.split_to(buf_length as usize)).await,
+                Err(negative) => match negative {
+                    // the query is valid but the Decider can not accommodate them.
+                    NegativeMultiDecision::BatchNonConforming(_, not_until) => {
+                        // We can either change buf_length or we can wait until the rate limiter has enough capacity to accommodate the request.
+                        let wait_time = not_until
```

Comment:
Convert to secs maybe to large? Can we use the duration returned here directly?

Reply:
Done